### ISSUE TITLE
[Issue]: RCCL All-reduce results depend on message size: algorithm-by-size selection changes reduction order (undocumented determinism hazard)

### ISSUE THREAD

**Issue by curvedinf** [ROCm/rocm-systems#11661]
### Summary

`dist.all_reduce` picks its collective algorithm/protocol as a function of
message size. Different algorithms reduce in different orders — and
multi-hop ring algorithms quantize partial sums to the tensor dtype at each
hop — so the same data reduced at different shapes gives bitwise different
results. Measured on 4× AMD Instinct MI100 (gfx908, ROCm 7.14, bf16):
all-reduce of identical data at (1664, 5120) vs (2048, 5120) differs in
**~34% of elements** (pure-torch repro below, no framework involved).

Any consumer whose collective sizes vary dynamically gets silent,
geometry-dependent nondeterminism. Engine example: vLLM TP serving with
chunked prefill — every layer's all-reduce size changes between chunks and
between prefill/decode, so logits depend on the chunk size
(`--max-num-batched-tokens`); measured: >1 nat prompt-logprob differences at
chunk boundaries, 2–17 nat outliers, degraded long-context output.

### Expected behavior

Either document that results are not bitwise-stable across message sizes
(and how to pin algorithms per deployment), or provide a supported
fixed-reduction-order mode per shape class. Today neither is stated
anywhere users can find.

### Reproduction (pure torch, ≥2 same-architecture GPUs)

[rccl_size_dependent_repro.py](https://github.com/user-attachments/files/32256124/rccl_size_dependent_repro.py)
```python
# rccl_size_dependent_repro.py
import os

import torch
import torch.distributed as dist


def main() -> None:
    dist.init_process_group("nccl")  # maps to RCCL on ROCm
    r = dist.get_rank()
    local_rank = int(os.environ.get("LOCAL_RANK", 0))
    ndev = torch.cuda.device_count()
    torch.cuda.set_device(local_rank if local_rank < ndev else ndev - 1)
    g = torch.Generator(device="cuda").manual_seed(100 + r)
    H = 5120
    X = (torch.randn(2048, H, generator=g, device="cuda") * (0.1 + 0.05 * r)).to(
        torch.bfloat16
    )

    def ar(t):
        o = t.clone()
        dist.all_reduce(o)
        torch.cuda.synchronize()
        return o

    outB, outA = ar(X), ar(X[:1664].contiguous())
    d = (outA.float() - outB[:1664].float()).abs()
    print(
        f"[rank {r}] 1664-vs-2048: bitwise={torch.equal(outA, outB[:1664])} "
        f"frac_diff={(d > 0).float().mean().item():.5f} max={d.max():.3e}"
    )
    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

```bash
torchrun --nproc_per_node=4 rccl_size_dependent_repro.py
# -> [rank r] 1664-vs-2048: bitwise=False frac_diff ~= 0.34
```

### Ask

1. Document the behavior and the pinning knobs (`RCCL_ALGO` and friends) as
   a determinism hazard for variable-size workloads.
2. Consider a deterministic mode (fixed algorithm per shape class, or fp32
   partials with a single downcast at hop boundaries).

### Environment

4× AMD Instinct MI100 (gfx908, XGMI), ROCm 7.14, RCCL as shipped with torch
(wheel 2.11.0+rocm7.1). Repro script validated to the communicator boundary on
ROCm 7.0 / RCCL 2.26.6 (single-dGPU host; the measurement itself requires
≥2 same-arch GPUs).

### Operating System

Ubuntu 22.04

### CPU

Epyc 7K62

### GPU

4x MI100, XGMI

### ROCm Version

7.14

### Installation Method

Package manager

### (Optional for Linux users) Output of rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

**Comment by Ko496-glitch** [comment]
@curvedinf  i can take look at this
